### PR TITLE
Update filename construction to include accumulation period

### PR DESCRIPTION
**Pull request by gavinevans**
Addresses part of #663 

Description
As part of #663, this adds support for a filename to be generated that contains an accumulation period. The resulting filename when it is specified that a period will be included will contain an additional period identifier of the format "PT01H" or "PT15M".

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/tests/utilities/test_filename.py [hunk 1]
```diff
@@ -47,69 +46,127 @@ class Test_generate_file_name(IrisTest):
...
+        """Test including a period within the filename when the period is in
+        minutes and deduced from the time coordinate."""
+        self.cube15m.coord("forecast_period").bounds = None
+        name = generate_file_name(self.cube15m, include_period=True)
+        self.assertIsInstance(name, str)
+        self.assertEqual(
+            name, "20151119T0030Z-PT0000H15M-air_temperature-PT15M.nc")
+
+    def test_no_bounds_exception(self):
+        """Test that an exception is raised forecast_period and time
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/tests/utilities/test_filename.py]: I've updated this docstring.

File: lib/improver/tests/utilities/test_filename.py [hunk 2]
```diff
@@ -47,69 +46,127 @@ class Test_generate_file_name(IrisTest):
...
+        coordinates provided do not have bounds."""
+        self.cube1h15m.coord("forecast_period").bounds = None
+        self.cube1h15m.coord("time").bounds = None
+        msg = "Neither the forecast_period coordinate"
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/tests/utilities/test_filename.py]: This is just a partial error message that should match against this [full error message](https://github.com/metoppv/improver/blob/eff722f38bdc2add570ea1237ed0b55724bf560c/lib/improver/utilities/filename.py#L103).  For the unit test, I've only included a test for part of the error message, as hopefully this is distinctive enough.

File: lib/improver/utilities/filename.py [hunk 3]
```diff
@@ -81,7 +89,48 @@ def generate_file_name(cube, parameter=None):
             parameter = parameter.replace(char, '')
         parameter = parameter.replace('__', '_')
 
-    filename = '{}-{}-{}.nc'.format(
-        validity_time_string, forecast_period_string, parameter)
+    period_string = None
+    if include_period:
+        coord_units = {"forecast_period": "seconds",
+                       "time": "seconds since 1970-01-01 00:00:00"}
+        for coord_name in ["forecast_period", "time"]:
```
Comment (fionaRust) [inline lib/improver/utilities/filename.py]: It is fairly clear what this is doing but it wouldn't hurt to put a comment above explaining it.

File: lib/improver/utilities/filename.py [hunk 4]
```diff
@@ -49,6 +54,9 @@ def generate_file_name(cube, parameter=None):
         parameter (str):
             Optional parameter name to use in the output filename rather than
             taking the name of the cube diagnostic.
+        include_period (bool):
+            Keyword argument to indicate whether a period, accumulation or
+            time window identifier should be included within the filename.
 
     Returns:
         filename (str):
```
Comment (fionaRust) [inline lib/improver/utilities/filename.py]: You need to update your Raises section too.

File: lib/improver/utilities/filename.py (hunk 3 above)
Comment (gavinevans) [reply lib/improver/utilities/filename.py]: Comment added.

File: lib/improver/utilities/filename.py (hunk 4 above)
Comment (gavinevans) [reply lib/improver/utilities/filename.py]: Updated Raises section.

**Pull request merged**